File: code-generator/src/main/java/io/github/potjerodekool/openapi/internal/di/bean/LazySingletonScopeManager.java
package io.github.potjerodekool.openapi.internal.di.bean;

import org.checkerframework.checker.nullness.qual.Nullable;

public class LazySingletonScopeManager<T> implements ScopeManager<T> {

    private final BeanDefinition beanDefinition;
    private @Nullable T instance = null;
    private boolean isInit = false;

    public LazySingletonScopeManager(final BeanDefinition beanDefinition) {
        this.beanDefinition = beanDefinition;
    }

    @Override
    public T get(final DefaultApplicationContext applicationContext) {
        synchronized (this) {
            if (!isInit) {
                isInit = true;
                instance = applicationContext.createBean(beanDefinition);
            }
        }

        if (instance == null) {
            throw new IllegalStateException(String.format("Failed to get instance of %s", getBeanType().getName()));
        }

        return instance;
    }

    @Override
    public Class<T> getBeanType() {
        return (Class<T>) beanDefinition.getBeanType();
    }
}

File: code-generator/src/main/kotlin/io/github/potjerodekool/openapi/internal/di/bean/LazySingletonScopeManager.kt
package io.github.potjerodekool.openapi.internal.di.bean

class LazySingletonScopeManager<T>(private val beanDefinition: BeanDefinition) : ScopeManager<T> {
    private var instance: T? = null
    private var isInit = false

    override fun get(applicationContext: DefaultApplicationContext?): T {
        synchronized(this) {
            if (!isInit) {
                isInit = true
                instance = applicationContext!!.createBean(beanDefinition)
            }
        }

        checkNotNull(instance) { String.format("Failed to get instance of %s", beanType!!.name) }

        return instance as (T & Any)
    }

    override val beanType: Class<T>?
        get() {
            val beanType = beanDefinition.beanType ?: return null
            return beanType.java as Class<T>
        }

    /*
    override val beanType: Class<T>?
        get() = {
            if (beanDefinition.beanType != null) beanDefinition.beanType::class.java as Class<T>?
            else null
        }
            */
}
